modules/_node-scripts/tsc/runTscChecks.mjs

```diff
@@ -45,36 +46,23 @@ export default async function runTscChecks(modifiedSince) {
 			console.log(`ℹ️ Going to check ${projectDirs.length} projects`);
 		}
 
-		const projectGroups = [[]];
+		const asyncTaskQueue = new AsyncTaskQueue();
 
 		for (const projectDir of projectDirs) {
 			if (!(await fileExists(path.join(projectDir, SRC_TSCONFIG_PATH)))) {
 				continue;
 			}
 
-			let group = projectGroups[projectGroups.length - 1];
-
-			if (group.length === cpuCount) {
-				group = [];
-
-				projectGroups.push(group);
-			}
-
-			group.push(projectDir);
-		}
-
-		for (const projectGroup of projectGroups) {
-			const all = await Promise.all(
-				projectGroup.map((projectDir) => {
+			asyncTaskQueue.runTask(
+				async () => {
 					console.log(
 						`🕵️ Checking ${path.relative(rootDir, projectDir)}`
 					);
 
-					return runTsc(projectDir);
-				})
+					return await runTsc(projectDir);
+				},
+				(result) => results.push(result)
```

Comment:
Does this callback based approach work? Because once you add the last task if the queue is still full, it will finish immediately which will cause `runTscChecks` to finish with an incomplete `result` array.

Also, since we are using promises it would be better if AsyncTaskQueue implemented a Promise based contract.

Maybe, to make it simpler, you could just define a method to add all tasks at once with a single array (instead of adding them one by one) and handle all the synchronization there. Then that method would return a Promise with all the results.

Alternatively, you could make runTask return a Promise that fulfills when the task is finished, instead of having to pass a callback. That way you fill an array of promises for all tasks in the for loop, then await on Promise.all(array) after the loop...